concurrency-limit low-priority initial logical size calculation [v2]

problame opened the pull request:
Problem
-------

Before this PR, there was no concurrency limit on initial logical size computations.

While logical size computations are lazy in theory, in practice (production), they happen in a short timeframe after restart.

This means that on a PS with 20k tenants, we'd have up to 20k concurrent initial logical size calculation requests.

This is self-inflicted needless overload.

This hasn't been a problem so far because the `.await` points on the logical size calculation path never return `Pending`, hence we have a natural concurrency limit of the number of executor threads.
But, as soon as we return `Pending` somewhere in the logical size calculation path, other concurrent tasks get scheduled by tokio.
If these other tasks are also logical size calculations, they eventually pound on the same bottleneck.

For example, in #5479, we want to switch the VirtualFile descriptor cache to a `tokio::sync::RwLock`, which makes us return `Pending`, and without measures like this patch, after PS restart, VirtualFile descriptor cache thrashes heavily for 2 hours until all the logical size calculations have been computed and the degree of concurrency / concurrent VirtualFile operations is down to regular levels.
See the *Experiment* section below for details.

<!-- Experiments (see below) show that plain #5479 causes heavy thrashing of the VirtualFile descriptor cache.
The high degree of concurrency is too much for 
In the case of #5479 the VirtualFile descriptor cache size starts thrashing heavily.


-->

Background
----------

Before this PR, initial logical size calculation was spawned lazily on first call to `Timeline::get_current_logical_size()`.

In practice (prod), the lazy calculation is triggered by `WalReceiverConnectionHandler` if the timeline is active according to storage broker, or by the first iteration of consumption metrics worker after restart (`MetricsCollection`).

The spawns by walreceiver are high-priority because logical size is needed by Safekeepers (via walreceiver `PageserverFeedback`) to enforce the project logical size limit.
The spawns by metrics collection are not on the user-critical path and hence low-priority. [^consumption_metrics_slo]

[^consumption_metrics_slo]: We can't delay metrics collection indefintely because there are TBD internal SLOs tied to metrics collection happening in a timeline manner (https://github.com/neondatabase/cloud/issues/7408). But let's ignore that in this issue.

The ratio of walreceiver-initiated spawns vs consumption-metrics-initiated spawns can be reconstructed from logs (`spawning logical size computation from context of task kind {:?}"`).
PR #5995 and #6018 adds metrics for this.

First investigation of the ratio lead to the discovery that walreceiver spawns 75% of init logical size computations.
That's because of two bugs:
- In Safekeepers: https://github.com/neondatabase/neon/issues/5993
- In interaction between Pageservers and Safekeepers: https://github.com/neondatabase/neon/issues/5962

The safekeeper bug is likely primarily responsible but we don't have the data yet. The metrics will hopefully provide some insights.

When assessing production-readiness of this PR, please assume that neither of these bugs are fixed yet.


Changes In This PR
------------------

With this PR, initial logical size calculation is reworked as follows:

First, all initial logical size calculation task_mgr tasks are started early, as part of timeline activation, and run a retry loop with long back-off until success. This removes the lazy computation; it was needless complexity because in practice, we compute all logical sizes anyways, because consumption metrics collects it.

Second, within the initial logical size calculation task, each attempt queues behind the background loop concurrency limiter semaphore. This fixes the performance issue that we pointed out in the "Problem" section earlier.

Third, there is a twist to queuing behind the background loop concurrency limiter semaphore. Logical size is needed by Safekeepers (via walreceiver `PageserverFeedback`) to enforce the project logical size limit. However, we currently do open walreceiver connections even before we have an exact logical size. That's bad, and I'll build on top of this PR to fix that (https://github.com/neondatabase/neon/issues/5963). But, for the purposes of this PR, we don't want to introduce a regression, i.e., we don't want to provide an exact value later than before this PR. The solution is to introduce a priority-boosting mechanism (`GetLogicalSizePriority`), allowing callers of `Timeline::get_current_logical_size` to specify how urgently they need an exact value. The effect of specifying high urgency is that the initial logical size calculation task for the timeline will skip the concurrency limiting semaphore. This should yield effectively the same behavior as we had before this PR with lazy spawning.

Last, the priority-boosting mechanism obsoletes the `init_order`'s grace period for initial logical size calculations. It's a separate commit to reduce the churn during review. We can drop that commit if people think it's too much churn, and commit it later once we know this PR here worked as intended.

Experiment With #5479 
---------------------

I validated this PR combined with #5479 to assess whether we're making forward progress towards asyncification.

The setup is an `i3en.3xlarge` instance with 20k tenants, each with one timeline that has 9 layers. 
All tenants are inactive, i.e., not known to SKs nor storage broker.
This means all initial logical size calculations are spawned by consumption metrics `MetricsCollection` task kind.
The consumption metrics worker starts requesting logical sizes at low priority immediately after restart. This is achieved by deleting the consumption metrics cache file on disk before starting PS.[^consumption_metrics_cache_file]

[^consumption_metrics_cache_file] Consumption metrics worker persists its interval across restarts to achieve persistent reporting intervals across PS restarts; delete the state file on disk to get predictable (and I believe worst-case in terms of concurrency during PS restart) behavior.

Before this patch, all of these timelines would all do their initial logical size calculation in parallel, leading to extreme thrashing in page cache and virtual file cache.

With this patch, the virtual file cache thrashing is reduced significantly (from 80k `open`-system-calls/second to ~500 `open`-system-calls/second during loading).


### Critique

The obvious critique with above experiment is that there's no skipping of the semaphore, i.e., the priority-boosting aspect of this PR is not exercised.

If even just 1% of our 20k tenants in the setup were active in SK/storage_broker, then 200 logical size calculations would skip the limiting semaphore immediately after restart and run concurrently.

Further critique: given the two bugs wrt timeline inactive vs active state that were mentioned in the Background section, we could have 75% of our 20k tenants being (falsely) active on restart.

So... (next section)

This Doesn't Make Us Ready For Async VirtualFile
------------------------------------------------

This PR is a step towards asynchronous `VirtualFile`, aka, #5479 or even #4744.

But it doesn't yet enable us to ship #5479.

The reason is that this PR doesn't limit the amount of high-priority logical size computations.
If there are many high-priority logical size calculations requested, we'll fall over like we did if #5479 is applied without this PR.
And currently, at very least due to the bugs mentioned in the Background section, we run thousands of high-priority logical size calculations on PS startup in prod.

So, at a minimum, we need to fix these bugs.

Then we can ship #5479 and #4744, and things will likely be fine under normal operation.

But in high-traffic situations, overload problems will still be more likely to happen, e.g., VirtualFile cache descriptor thrashing.
The solution candidates for that are orthogonal to this PR though:
* global concurrency limiting
* per-tenant rate limiting => #5899
* load shedding
* scaling bottleneck resources (fd cache size (neondatabase/cloud#8351), page cache size(neondatabase/cloud#8351), spread load across more PSes, etc)

Conclusion
----------

Even with the remarks from in the previous section, we should merge this PR because:
1. it's an improvement over the status quo (esp. if the aforementioned bugs wrt timeline active / inactive are fixed)
2. it prepares the way for https://github.com/neondatabase/neon/pull/6010
3. it gets us close to shipping #5479 and #4744

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1759,39 +1725,60 @@ impl Timeline {
+                // (The alternative would be to calculate the size here, but,
+                //  it can actually take a long time if the user has a lot of rels.
+                //  And we'll inevitable need it again; So, let the background task do the work.)
+                match self
+                    .current_logical_size
+                    .cancel_wait_for_background_loop_concurrency_limit_semaphore
+                    .get()
+                {
+                    Some(cancel) => cancel.cancel(),
+                    None => {
+                        warn!("unexpected: priority_tx not set, logical size calculation will not be prioritized");
+                    }
+                };
```
This is the priority boosting part.

problame commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1808,89 +1793,153 @@ impl Timeline {
-                Ok(())
-            }.in_current_span(),
-        );
+        // set the gauge value before it can be set in `update_current_logical_size`.
+        // TODO: shouldn't this simple .add(calculated_size)?
+        self.metrics.current_logical_size_gauge.set(sum);
```
The TODO is pointing out that we're losing the `update_current_logical_size()` deltas in the window from the point when we `.load()` the `added` to the point where we `initial_logical_size.set()` below.

problame commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1928,6 +1977,7 @@ impl Timeline {
         receiver
     }
 
+    /// TODO: must be cancellation safe, I think it is (?)
```
TODO validate this.

problame replied:
Waiting for https://github.com/neondatabase/neon/pull/6031

problame commented:
@koivunej I polished up the PR description, re-ran the experiment to make sure this is a step forward with regard to #5479 , and added two sections explaining what exactly that means.

The remaining comment is to make sure that logical size computation impl is cancel safe: https://github.com/neondatabase/neon/pull/6000#discussion_r1412049728 

I think it is, and added a commit that adds cancel-safety comments to the methods that I audited. I'm treating the `Layer::get_reconstruct_data` as a black box.

Would appreciate you to comment there, and resolve your other conversation.

Then let's get this shipped so we maximize soak time.

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1797,39 +1748,88 @@ impl Timeline {
-        // We need to start the computation task.
-        // It gets a separate context since it will outlive the request that called this function.
+        let cancel_wait_for_background_loop_concurrency_limit_semaphore = CancellationToken::new();
+        self.current_logical_size.cancel_wait_for_background_loop_concurrency_limit_semaphore.set(cancel_wait_for_background_loop_concurrency_limit_semaphore.clone()).expect("initial logical size calculation task must be spawned exactly once per Timeline object");
```
```suggestion
        let token = cancel_wait_for_background_loop_concurrency_limit_semaphore.clone();
        self.current_logical_size
            .cancel_wait_for_background_loop_concurrency_limit_semaphore.set(token)
            .expect("initial logical size calculation task must be spawned exactly once per Timeline object");
```

Let's help rustfmt out.

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1844,96 +1844,152 @@ impl Timeline {
-                        // because initial_logical_size is still None.
-                        info!("initial size calculation cancelled, likely timeline delete / tenant detach");
-                        return Ok(());
+                        Err(BackgroundCalculationError::Cancelled)
```
Even after this PR were the only cancellations we expect from shutdowns or deletions?

I am just wondering about a metric for this. But probably it is not needed, I should just fix the download path completing the initialization for Layer anyways or at least support evicting downloaded-but-not-initialized Layers, which cancellation would bring.
